sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/numberFormatExpressions.scala

```diff
@@ -22,48 +22,61 @@ import java.util.Locale
 import org.apache.spark.sql.catalyst.analysis.TypeCheckResult
 import org.apache.spark.sql.catalyst.expressions.codegen.{CodegenContext, CodeGenerator, ExprCode}
 import org.apache.spark.sql.catalyst.expressions.codegen.Block.BlockHelper
-import org.apache.spark.sql.catalyst.util.NumberFormatter
+import org.apache.spark.sql.catalyst.util.ToNumberParser
 import org.apache.spark.sql.types.{DataType, StringType}
 import org.apache.spark.unsafe.types.UTF8String
 
 /**
- * A function that converts string to numeric.
+ * A function that converts strings to decimal values, returning an exception if the input string
+ * fails to match the format string.
  */
 @ExpressionDescription(
   usage = """
-     _FUNC_(strExpr, formatExpr) - Convert `strExpr` to a number based on the `formatExpr`.
-       The format can consist of the following characters:
-         '0' or '9':  digit position
-         '.' or 'D':  decimal point (only allowed once)
-         ',' or 'G':  group (thousands) separator
-         '-' or 'S':  sign anchored to number (only allowed once)
-         '$':  value with a leading dollar sign (only allowed once)
+     _FUNC_(expr, fmt) - Convert string 'expr' to a number based on the string format 'fmt'.
+       Throws an exception if the conversion fails. The format can consist of the following
+       characters, case insensitive:
+         '0' or '9':  Specifies an expected digit between 0 and 9. A 0 to the left of the decimal
+           point indicates that 'expr' must have at least as many digits. A leading 9 indicates
+           that 'expr' may omit these digits. For any combination of digits 0 and/or 9, the behavior
+           is decided by the left-most digit. 'expr' must not be larger than the number of digits
+           allowed to the left of the decimal point per the format string. Digits in the format
+           string to the right of the decimal indicate the most digits 'expr' may have to the right
+           of the decimal point.
+         '.' or 'D': Specifies the position of the decimal point (optional, only allowed once).
+         ',' or 'G': Specifies the position of the grouping (thousands) separator (,). There must be
+           one or more 0 or 9 to the left of the rightmost grouping separator. 'expr' must match the
+           grouping separator relevant for the size of the number.
+         '$': Specifies the location of the $ currency sign. This character may only be specified
+           once.
+         'S': Specifies the position of a '+' or '-' sign (optional, only allowed once).
+         'MI': Only allowed at the end of the format string; specifies that 'expr' has an optional
```

Comment:
looks like snowflake can put `MI` anywhere: https://docs.snowflake.com/en/sql-reference/sql-format-models.html

Reply:
seems like `MI` and `S` have no difference when parsing, the only difference is when printing (`to_char` function), `MI` prints a space for positive numbers, while `S` prints `+`.

Reply:
Sure, that's just updating this condition:

```
    // Make sure the format string contains at most one minus sign at the end.
    else if (inputTokenCounts.getOrElse(OptionalMinusSign(), 0) > 1 ||
      (inputTokenCounts.getOrElse(OptionalMinusSign(), 0) == 1 &&
        formatTokens.last != OptionalMinusSign())) {
      notAtEndOfNumberFormatError(s"'$OPTIONAL_MINUS_STRING'")
    }
```

I updated it to remove that constraint.